adoc/chapters/programming_interface.adoc

```diff
@@ -9956,11 +9958,27 @@ The characteristics of USM allocations are summarized in
 |                                 Another [code]#device#    | Optional  | Another [code]#device#| Optional
 |====
 
-USM allocations are only guaranteed to be valid in the SYCL context in which
-they were created. The behavior of accessing a USM allocation in a different
-context from which it was created is undefined.  Attempting to access a kind
-of USM allocation that is not supported by the device results in undefined
-behavior.
+Each USM memory allocation has an associated SYCL context, and any access to
+that memory must use the same context.  Specifically, any
+<<sycl-kernel-function>> that dereferences a pointer to a USM memory allocation
+must be submitted to a <<queue>> that was constructed with the same context
+that was used to allocate that USM memory.  The explicit memory operation
+<<command, commands>> that take USM pointers have a similar restriction.  (See
+<<subsec:explicitmemory>> for details.)  Violations of these requirements
+result in undefined behavior.
+
+[NOTE]
+====
+There are no similar restrictions for dereferencing a USM pointer in a
+<<host-task>>.  This is legal regardless of which <<queue>> the host task was
+submitted to so long as the USM pointer is accessible on the host.
+====
+
+Each type of USM memory allocation has different rules for where that memory is
+accessible.  Attempting to dereference a USM pointer in violation of these
+rules results in undefined behavior.  The explicit memory operation
+<<command, commands>> that take USM pointers have a similar restriction, and
+violating these restrictions results in undefined behavior.
```

Comment:
I assume the last sentence in your suggestion is supposed to mean that it is UB to pass a USM pointer to one of the "explicit memory operation" functions unless that pointer is accessible on the device?  This doesn't seem very clear, though, because a reader might not think of these functions as "accessing" the memory at that pointer.

How about if the last sentence says:

> Passing a USM pointer to one of the explicit memory functions where the pointer is not accessible to the device generally results in undefined behavior.  See <<subsec:explicitmemory>> for the exact rules.

I wanted to be a bit vague here to make it easier in the future to allow cases where (e.g.) the application passes a USM host pointer which is not accessible to the device.  I want the description of the functions in section `<<subsec:explicitmemory>>` to be the single place where these rules are defined.

Note that this paragraph differs from the previous one because the previous paragraph talks about the rules regarding "context".  By contrast, this paragraph talks about the rules for "accessibility" of the pointer.  In fact, I see this entire paragraph as the replacement of the old sentence "Attempting to access a kind of USM allocation that is not supported by the device results in undefined behavior" because it clarifies what "access" means.  I'm saying it means that you cannot dereference the pointer in a kernel and you cannot pass the pointer to the explicit memory operation functions.

Reply:
Just to be clear, the sentence from the old description that seems to be missing is:

> Attempting to access a kind of USM allocation that is not supported by the device results in undefined behavior.

I read this as saying: If you try to use a type of USM (say, shared USM) and the device does not support that type of USM (say, the device does not have aspect::usm_shared_allocations) then it won't work.  I would actually hope this is an error and not undefined behavior, but maybe I'm missing some nuance.

I tried to keep the previous sentence more-or-less unchanged but I'm fine rephrasing it or adding more detail.  Note though, I do think this statement should apply to all USM accesses, and not just the "explicit memory functions".

Reply:
> I would actually hope this is an error and not undefined behavior, but maybe I'm missing some nuance.

I would also like the "explicit memory operation" functions to return an error if the USM pointer has the wrong context or is not accessible on the device.  However, I did not want to write this into the spec if it results in significant overhead.  Do the underlying OpenCL and Level Zero "copy" APIs return an error in these cases?  I think their specs do not specifically say whether this is the case.

In general, I've been leaning towards UB for error conditions in these "clarification" PRs when I'm not sure about the cost of error checking.  I figure it's easier to change the spec later to say that an error must be returned, if we decide that it is reasonable.

> Note though, I do think this statement should apply to all USM accesses, and not just the "explicit memory functions".

Sure, 100% agree.  I think there are only two ways to use a USM pointer: you can dereference it in code, or you can pass it to one of the "explicit memory operation" functions.  I tired to address the dereferencing case with this sentence:

> Attempting to dereference a USM pointer in violation of these rules results in undefined behavior.

Thinking about it more, maybe it should be clarified a bit to:

> Attempting to dereference a USM pointer on the host or on a device in violation of these rules results in undefined behavior.

Note that the actual rules for whether a USM pointer is "accessible" on a device are not clear in this PR (or before this PR).  These are the rules in #8 that we are still debating.

Reply:
This paragraph updated in 691f960148e7d7a745cd9b6ee6b449852b68c4fe